experimental/https_attestation/src/server.rs

```diff
@@ -0,0 +1,239 @@
+//
+// Copyright 2021 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use anyhow::{anyhow, Context};
+use futures::TryStreamExt;
+use futures_util::future::TryFutureExt;
+use http::Uri;
+use hyper::{service::service_fn, Body, Client, Request, Response};
+use log::warn;
+use oak_attestation_common::keying_material::{Assertion, KeyingMaterialBundle};
+use std::sync::{Arc, Mutex};
+use tokio::net::TcpListener;
+use tokio_rustls::TlsAcceptor;
+
+#[derive(Clone)]
+enum ConnectionStatus {
+    Attested,
+    NotAttested,
+}
+
+/// HTTPS proxy that terminates TLS connections and proxies corresponding HTTP requests to the
+/// `backend_uri`.
+/// It checks client assertions and also generates sends back server assertions in
+/// order to perform remote attestation, if receives an HTTP POST request on `/report`.
+/// A separate `Proxy` instance is created for each individual TLS session.
```

Comment:
IIUC this is an HTTP server that intercepts all incoming HTTP request, and intercepts those to the /report path and provides remote attestation to the client, and forwards all other requests to the backend server. Is this correct? If so, could you explain some of this in the comment?

Reply:
Updated